File: backend/app/domain/events.py
"""SSE event contract shared between backend modes and the frontend.

The frontend mirrors these types in src/types/events.ts.
Adding a new event type requires updating both sides.
"""

from typing import Any, Literal, TypedDict


class TokenEvent(TypedDict):
    type: Literal["token"]
    content: str


class DetectedLangEvent(TypedDict):
    type: Literal["detected_lang"]
    code: str
    display: str


class AgentStartEvent(TypedDict):
    type: Literal["agent_start"]
    agent: str
    icon: str
    message: str


class AgentActionEvent(TypedDict):
    """Mono-agent ReAct: the agent picks a tool to invoke."""
    type: Literal["agent_action"]
    step: int
    tool: str
    icon: str
    input: str


class AgentObservationEvent(TypedDict):
    """Mono-agent ReAct: the result returned by the tool."""
    type: Literal["agent_observation"]
    step: int
    tool: str
    result: str


class AgentDoneEvent(TypedDict):
    """Multi-agent: a specialized agent finished its step."""
    type: Literal["agent_done"]
    agent: str
    icon: str
    result: str
    metadata: dict[str, Any]


class FinalEvent(TypedDict):
    type: Literal["final"]
    translation: str
    metadata: dict[str, Any]


class ErrorEvent(TypedDict):
    type: Literal["error"]
    message: str

File: frontend/src/types/events.ts
// Mirror of backend/app/domain/events.py
// Adding a new event type requires updating both files.

export type ModeId = 'pipeline' | 'monoagent' | 'multiagent';

export interface ModeInfo {
  id: ModeId;
  name: string;
  description: string;
  streams_tokens: boolean;
  streams_steps: boolean;
}

export interface TokenEvent {
  type: 'token';
  content: string;
}

export interface DetectedLangEvent {
  type: 'detected_lang';
  code: string;
  display: string;
}

export interface AgentStartEvent {
  type: 'agent_start';
  agent: string;
  icon: string;
  message: string;
}

export interface AgentActionEvent {
  type: 'agent_action';
  step: number;
  tool: string;
  icon: string;
  input: string;
}

export interface AgentObservationEvent {
  type: 'agent_observation';
  step: number;
  tool: string;
  result: string;
}

export interface AgentDoneEvent {
  type: 'agent_done';
  agent: string;
  icon: string;
  result: string;
  metadata: Record<string, unknown>;
}

export interface FinalEvent {
  type: 'final';
  translation: string;
  metadata: Record<string, unknown>;
}

export interface ErrorEvent {
  type: 'error';
  message: string;
}

export type StreamEvent =
  | TokenEvent
  | DetectedLangEvent
  | AgentStartEvent
  | AgentActionEvent
  | AgentObservationEvent
  | AgentDoneEvent
  | FinalEvent
  | ErrorEvent;

export interface TranslateRequest {
  text: string;
  source_lang: string;
  target_lang: string;
}
